metrics/ci_integrations/lib/source/github_actions/adapter/github_actions_source_client_adapter.dart

```diff
@@ -0,0 +1,326 @@
+  ///
+  /// A [workflowIdentifier] is either a workflow id or a name of the file
+  /// that defines the workflow.
+  ///
+  /// A [status] is used as a filter query parameter to define the
+  /// [WorkflowRun.status] of runs to fetch.
+  ///
+  /// A [perPage] is used for limiting the number of runs and pagination in pair
+  /// with the [page] parameter.
+  ///
+  /// A [page] is used for pagination and defines a page of runs to fetch.
+  /// If the [page] is `null` or omitted, the first page is fetched.
+  Future<WorkflowRunsPage> _fetchRunsPage({
+    String workflowIdentifier,
+    RunStatus status,
+    int perPage,
+    int page,
+  }) async {
+    final runsInteraction = await githubActionsClient.fetchWorkflowRuns(
+      workflowIdentifier,
+      status: status,
+      perPage: perPage,
+      page: page,
+    );
+
+    _throwIfInteractionUnsuccessful(runsInteraction);
+
+    return runsInteraction.result;
+  }
+
+  /// Fetches a [WorkflowRunArtifactsPage] of a run with the given [runId].
+  ///
+  /// A [perPage] is used for limiting the number of runs and pagination in pair
+  /// with the [page] parameter.
+  ///
+  /// A [page] is used for pagination and defines a page of runs to fetch.
+  /// If the [page] is `null` or omitted, the first page is fetched.
+  Future<WorkflowRunArtifactsPage> _fetchArtifactsPage({
+    int runId,
+    int perPage,
+    int page,
+  }) async {
+    final artifactsInteraction = await githubActionsClient.fetchRunArtifacts(
+      runId,
+      perPage: perPage,
+      page: page,
+    );
+
+    _throwIfInteractionUnsuccessful(artifactsInteraction);
+
+    return artifactsInteraction.result;
+  }
+
+  /// Fetches the next [WorkflowRunsPage] of the given [currentPage].
+  Future<WorkflowRunsPage> _fetchNextRunsPage(
+    WorkflowRunsPage currentPage,
+  ) async {
+    final pageInteraction =
+        await githubActionsClient.fetchNextRunsPage(currentPage);
+
+    _throwIfInteractionUnsuccessful(pageInteraction);
+
+    return pageInteraction.result;
+  }
+
+  /// Fetches the next [WorkflowRunArtifactsPage] of the given [currentPage].
+  Future<WorkflowRunArtifactsPage> _fetchNextArtifactsPage(
+    WorkflowRunArtifactsPage currentPage,
+  ) async {
+    final pageInteraction =
+        await githubActionsClient.fetchNextRunArtifactsPage(currentPage);
+
+    _throwIfInteractionUnsuccessful(pageInteraction);
+
+    return pageInteraction.result;
+  }
+
+  /// Fetches the given [artifact], searches for the coverage file and converts
+  /// it to [Coverage].
+  ///
+  /// If the coverage file was not found, returns `null`.
+  Future<Percent> _mapArtifactToCoverage(
+    WorkflowRunArtifact artifact,
+  ) async {
+    final artifactInteraction =
+        await githubActionsClient.downloadRunArtifactZip(artifact.downloadUrl);
+    _throwIfInteractionUnsuccessful(artifactInteraction);
+
+    final artifactBytes = artifactInteraction.result;
+    final coverageJsonFile = ArchiveUtil.getArchiveFile(
```

Comment:
Once the ArchiveUtil is refactored, should we integrate the resulting class to ensure this adapter is testable enough and doesn't use the File System in test cases?